### ISSUE TITLE
Permanent 100% CPU usage minutes after LSP restart in SublimeText, using ruby-lsp, ruby-lsp-rails

### ISSUE THREAD

**Issue by maxim** [Shopify/ruby-lsp#1306]
### Ruby version

3.2.2

### Description

I've been using latest ruby-lsp on latest Sublime Text for a while. It has always had an issue where the process would just sit at 100% CPU continuously after some time, and I would kill it, and let it restart.

After this week's series of updates however (I've been updating every day with hopes to get fixes), something went a bit haywire.

1. The ruby-lsp process starts using 100% cpu very soon after restart (after indexing finished, so this is not the indexing process)
2. Sublime starts showing weird Prism errors that don't actually exist.
3. Issues shown in Sublime get "stuck" (during that 100% cpu usage), editing code doesn't make them go away.

I'm sorry I didn't have a chance to collect more debugging info, but if someone with more knowledge could test this lsp in Sublime, would much appreciate.

Here are my LSP settings in sublime:

```json
{
  "clients":
  {
    "ruby-lsp": {
      "enabled": false,
      "command": ["ruby-lsp"],
      "selector": "source.ruby",
      "initializationOptions": {
        "experimentalFeaturesEnabled": true,
      },
    },
    "ruby": {
      "enabled": true,
      "command": ["solargraph", "stdio"],
      "selector": "source.ruby | text.html.ruby | text.html.rails",
    },
  },

  "diagnostics_highlight_style": {
    "error": "box",
    "warning": "box",
    "info": "box",
    "hint": "squiggly",
  },
}
```

### Expected output

Expected to see real issues, and no constant 100% cpu usage.

**Comment by maxim** [comment]
@vinistock Yes, you are spot on, I disabled ruby-lsp due to the aforementioned issues, and missed it when pasting the config. Will try to get some logs. 🤞🏻

**Comment by adam12** [comment]
I've been seeing this as well, but with Neovim. I am not sure if it's directly related. It only seems to have started happening recently (I first looked at it about 5 weeks ago).

Attaching `lldb`, and listing threads, the primary Ruby thread seems to be inside `server.rb` on line `121`. Comparing it to my installed version, it lines up with this code block:

https://github.com/Shopify/ruby-lsp/blob/1159866dcce6b3619319df99806bcc03a46f808d/lib/ruby_lsp/server.rb#L120-L140

I assume that `pop` isn't blocking as we'd expect here, and it's caused a busy loop.

I haven't been able to get a good backtrace. Selecting the thread in `lldb` and then calling `call (void) rb_backtrace()` appears to work, but it's not being output into my LSP log. 

One thing to note, and this might just be anecdotal, is that the `ruby-lsp` process I currently have consuming 100% has _no_ parent nvim process. It disappeared at some point and either the ruby-lsp process wasn't notified to shutdown, or it was and didn't.

**Comment by vinistock** [comment]
The `start` method is the main LSP loop. It never exits unless it receives a `shutdown` and then an `exit` request.

Also, it's expected for the Thread worker to show up in profiles, since most requests are executed by that worker.

We did discover a performance issue with signature help, which should be solved by https://github.com/Shopify/ruby-lsp/pull/1369 included in [v0.14](https://github.com/Shopify/ruby-lsp/releases/tag/v0.14.0). Can you please give v0.14 a try? Also, if you're using `ruby-lsp-rails`, we're working on addressing the breaking changes so that the LSP can update.

**Comment by vinistock** [comment]
In that case, I'm closing the issue for now. If you encounter problems again, please re-open.
